io-sim/src/Control/Monad/IOSim.hs

```diff
@@ -266,3 +332,138 @@ ppEvents events =
+                                 controlSimTrace (explorationStepTimelimit opts) control mainAction
+          (sleeper,trace) = compareTraces passingTrace trace0
+      in (counterexample ("Schedule control: " ++ show control) $
+          counterexample (case sleeper of Nothing -> "No thread delayed"
+                                          Just ((t,tid,lab),racing) ->
+                                            showThread (tid,lab) ++
+                                            " delayed at time "++
+                                            show t ++
+                                            "\n  until after:\n" ++
+                                            unlines (map (("    "++).showThread) $ Set.toList racing)
+                                            ) $
+          k passingTrace trace) .&&|
+         let limit     = (n+m-1) `div` m
+             -- To ensure the set of schedules explored is deterministic, we filter out
+             -- cached ones *after* selecting the children of this node.
+             races     = filter (not . cached) . take limit $ readRaces()
+             branching = length races
+         in -- tabulate "Races explored" (map show races) $
+            tabulate "Branching factor" [bucket branching] $
+            tabulate "Race reversals per schedule" [bucket (raceReversals control)] $
+            conjoinPar
+              [ --Debug.trace "New schedule:" $
+                --Debug.trace ("  "++show r) $
+                --counterexample ("Schedule control: " ++ show r) $
+                explore n' ((m-1) `max` 1) r (Just trace0)
+              | (r,n') <- zip races (divide (n-branching) branching) ]
+
+    bucket n | n<10  = show n
+             | n>=10 = buck n 1
+             | otherwise = error "Ord Int is not a total order!"  -- GHC made me do it!
+    buck n t | n<10      = show (n*t) ++ "-" ++ show ((n+1)*t-1)
+             | n>=10     = buck (n `div` 10) (t*10)
+             | otherwise = error "Ord Int is not a total order!"  -- GHC made me do it!
+
+    divide n k =
+      [ n `div` k + if i<n `mod` k then 1 else 0
+      | i <- [0..k-1] ]
+
+    showThread :: (ThreadId,Maybe ThreadLabel) -> String
+    showThread (tid,lab) =
+      show tid ++ (case lab of Nothing -> ""
+                               Just l  -> " ("++l++")")
+
+    -- It is possible for the same control to be generated several times.
+    -- To avoid exploring them twice, we keep a cache of explored schedules.
+    cache = unsafePerformIO $ newIORef $
+              -- we use opts here just to be sure the reference cannot be
+              -- lifted out of exploreSimTrace
+              if explorationScheduleBound opts>=0
+                then Set.empty
+                else error "exploreSimTrace: negative schedule bound"
+    cached m = unsafePerformIO $ atomicModifyIORef' cache $ \set ->
+      (Set.insert m set, Set.member m set)
+    cacheSize () = unsafePerformIO $ Set.size <$> readIORef cache
```

Comment:
That would probably work, with the emphasis on probably. Because cacheSize uses unsafePerformIO, it may return different values at different times. And indeed, that's its purpose: to tell us the size of the cache at the moment it is called. That's why it makes sense for it to be a function. Removing the () parameter would make cacheSIze the size of the cache when it is _first_ evaluated. At the moment, I think it is called only once, so that ought to be OK... provided nobody inserts a second use in the future. Also, it's critically important that it is called at the right time, so we must rely on GHC's optimiser not moving the call, now or in the future. GHC's strictness analyser _does_ move evaluations of variables; if we replaced this by a variable, then I would be afraid that... at some point... GHC might evaluate it in a different place, and the code would stop working.

You might say that adding a () parameter is no absolute guarantee that GHC won't break the code, and I agree, but I think it's much less likely.